File: linked-list.cpp
#include <iostream>
#include <chrono>
using namespace std;
using namespace chrono;

class Node {
public:
    int data;
    Node* next;

    Node(int val) : data(val), next(nullptr) {}
};

class LinkedList {
private:
    Node* head;

public:
    LinkedList() : head(nullptr) {}

    // Insert at head
    void insertHead(int val) {
        Node* newNode = new Node(val);
        newNode->next = head;
        head = newNode;
    }

    // Insert at tail
    void insertTail(int val) {
        Node* newNode = new Node(val);
        if (!head) {
            head = newNode;
            return;
        }
        Node* temp = head;
        while (temp->next) {
            temp = temp->next;
        }
        temp->next = newNode;
    }

    // Delete a node
    void deleteNode(int val) {
        if (!head) return;
        if (head->data == val) {
            Node* toDelete = head;
            head = head->next;
            delete toDelete;
            return;
        }
        Node* temp = head;
        while (temp->next && temp->next->data != val) {
            temp = temp->next;
        }
        if (temp->next) {
            Node* toDelete = temp->next;
            temp->next = temp->next->next;
            delete toDelete;
        }
    }

    // Search for a node
    bool search(int val) {
        Node* temp = head;
        while (temp) {
            if (temp->data == val) return true;
            temp = temp->next;
        }
        return false;
    }

    // Traverse the list
    void traverse() {
        Node* temp = head;
        while (temp) {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }
};

int main() {
    LinkedList list;
    for (int i = 1; i <= 30000; i++) {
        list.insertHead(i);
    }
    /*
    auto start = high_resolution_clock::now();
    cout << list.search(5000) << endl;
    auto end = high_resolution_clock::now();
    cout << "Time taken for search: " << duration_cast<nanoseconds>(end - start).count() << " nanoseconds" << endl;
    */
    auto start = high_resolution_clock::now();
    cout << list.search(5000) << endl;
    auto end = high_resolution_clock::now();
    cout << "Time taken for search: " << duration_cast<seconds>(end - start).count() << " seconds" << endl;
    
    start = high_resolution_clock::now();
    list.traverse();
    end = high_resolution_clock::now();
    cout << "Time taken for traversal: " << duration_cast<seconds>(end - start).count() << " seconds" << endl;
    
    return 0;
}

File: linked-list.hs

import Data.Time.Clock (getCurrentTime, diffUTCTime)
import Text.Printf (printf)

-- Define the Node and LinkedList data structures
data Node = Node Int (Maybe Node)
newtype LinkedList = LinkedList (Maybe Node)

-- Create an empty list
emptyList :: LinkedList
emptyList = LinkedList Nothing

-- Insert at the head
insertHead :: Int -> LinkedList -> LinkedList
insertHead val (LinkedList headNode) = LinkedList (Just (Node val headNode))

-- Search for a value in the list
search :: Int -> LinkedList -> Bool
search _ (LinkedList Nothing) = False
search val (LinkedList (Just (Node dataVal nextNode)))
  | val == dataVal = True
  | otherwise = search val (LinkedList nextNode)

-- Traverse the list and print elements
traverseList :: LinkedList -> IO ()
traverseList (LinkedList Nothing) = putStrLn "End of list"
traverseList (LinkedList (Just (Node val nextNode))) = do
    putStr (show val ++ " ")
    traverseList (LinkedList nextNode)

-- Example usage
main :: IO ()
main = do
    -- Create a linked list and insert 10000 elements
    let list = foldr insertHead emptyList [1..30000]

    -- Measure time to search for an element
    startSearch <- getCurrentTime
    print $ search 5000 list
    endSearch <- getCurrentTime
    printf "Time taken for search: %.6f seconds\n" (realToFrac (diffUTCTime endSearch startSearch) :: Double)

    -- Measure time to traverse the list
    startTraverse <- getCurrentTime
    traverseList list
    endTraverse <- getCurrentTime
    printf "Time taken for traversal: %.6f seconds\n" (realToFrac (diffUTCTime endTraverse startTraverse) :: Double)